File: Data Structures & Algorithms/implement-prefix-tree/submission-1.java
class TrieNode{
    TrieNode[] children = new TrieNode[26];
    boolean endOfWord = false;
}
class PrefixTree {
    private TrieNode node;
    public PrefixTree() {
        node = new TrieNode();
    }

    public void insert(String word) {
        TrieNode cur = node;
        for(char c: word.toCharArray()){
            if(cur.children[c-'a'] == null){
                cur.children[c-'a'] = new TrieNode();
            }
            cur = cur.children[c-'a'];
        }
        cur.endOfWord = true;
    }

    public boolean search(String word) {
        TrieNode cur = node;
        for(char c: word.toCharArray()){
            if(cur.children[c-'a']==null) return false;
            cur = cur.children[c-'a'];
        }
        return cur.endOfWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode cur = node;
        for(char c: prefix.toCharArray()){
            if(cur.children[c-'a']==null) return false;
            cur = cur.children[c-'a'];
        }
        return true;
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-1.py
class TrieNode:
    def __init__(self):
        self.children = [None]*26
        self.word = False
class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        cur = self.root
        for c in word:
            i = ord(c) - ord('a')
            if cur.children[i] == None:
                cur.children[i] = TrieNode()
            cur = cur.children[i]
        cur.word=True

    def search(self, word: str) -> bool:
        def dfs(j, root):
            cur = root
            for i in range(j, len(word)):
                idx = ord(word[i]) - ord('a')
                if word[i] != '.':
                    if cur.children[idx]==None: return False
                else:
                    for c in range(0, 26):
                        if cur.children[c]!=None:
                            if dfs(i+1, cur.children[c]):
                                return True
                    return False
                cur = cur.children[idx]
            return cur.word
        return dfs(0, self.root)

            
